File: hostel_management_system_backend/Services/Verification/HostelVerificationService.cs
using AutoMapper;
using hostel_management_system_backend.Exceptions;

public interface IHostelVerificationService
{
    Task<HostelVerificationRequestReadDto> RequestVerificationAsync(Guid hostelId, Guid ownerId, CancellationToken cancellationToken);
    Task<HostelVerificationRequestReadDto> ApproveVerificationAsync(Guid requestId, Guid adminId, string? adminNotes, CancellationToken cancellationToken);
    Task<HostelVerificationRequestReadDto> RejectVerificationAsync(Guid requestId, Guid adminId, string? adminNotes, CancellationToken cancellationToken);
    Task<List<HostelVerificationRequestReadDto>> GetForHostelAsync(Guid hostelId, Guid requesterId, bool isAdmin, CancellationToken cancellationToken);
}

public sealed class HostelVerificationService : IHostelVerificationService
{
    private readonly IHostelVerificationRepository _repo;
    private readonly IMapper _mapper;

    public HostelVerificationService(IHostelVerificationRepository repo, IMapper mapper)
    {
        _repo = repo;
        _mapper = mapper;
    }

    public async Task<HostelVerificationRequestReadDto> RequestVerificationAsync(Guid hostelId, Guid ownerId, CancellationToken cancellationToken)
    {
        var hostel = await _repo.GetHostelForOwnerAsNoTrackingAsync(hostelId, ownerId, cancellationToken);
        if (hostel is null)
        {
            throw new ForbiddenException("You can only request verification for your own hostel.", "hostel_owner_required");
        }

        var pending = await _repo.GetPendingRequestForHostelAsync(hostelId, cancellationToken);
        if (pending is not null)
        {
            throw new ConflictException("A pending verification request already exists.", "verification_pending_exists");
        }

        var request = new HostelVerificationRequest
        {
            HostelId = hostelId,
            RequestedByUserId = ownerId,
            Status = HostelVerificationStatus.Pending,
            AdminNotes = null,
            ReviewedByAdminId = null,
            ReviewedAt = null,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = null,
            IsDeleted = false
        };

        await _repo.AddVerificationRequestAsync(request, cancellationToken);

        var hostelForUpdate = await _repo.GetHostelForUpdateAsync(hostelId, cancellationToken)
            ?? throw new NotFoundException("Hostel not found.");
        hostelForUpdate.VerificationStatus = HostelVerificationStatus.Pending;
        hostelForUpdate.UpdatedAt = DateTime.UtcNow;

        await _repo.SaveChangesAsync(cancellationToken);

        return _mapper.Map<HostelVerificationRequestReadDto>(request);
    }

    public async Task<HostelVerificationRequestReadDto> ApproveVerificationAsync(Guid requestId, Guid adminId, string? adminNotes, CancellationToken cancellationToken)
    {
        var request = await _repo.GetRequestForUpdateAsync(requestId, cancellationToken)
            ?? throw new NotFoundException("Verification request not found.");

        if (request.Status != HostelVerificationStatus.Pending)
        {
            throw new BadRequestException("Only pending requests can be approved.", "verification_request_not_pending");
        }

        request.Status = HostelVerificationStatus.Approved;
        request.AdminNotes = string.IsNullOrWhiteSpace(adminNotes) ? null : adminNotes.Trim();
        request.ReviewedByAdminId = adminId;
        request.ReviewedAt = DateTime.UtcNow;
        request.UpdatedAt = DateTime.UtcNow;

        var hostel = await _repo.GetHostelForUpdateAsync(request.HostelId, cancellationToken)
            ?? throw new NotFoundException("Hostel not found.");

        hostel.IsVerified = true;
        hostel.VerifiedAt = DateTime.UtcNow;
        hostel.VerifiedByAdminId = adminId;
        hostel.VerificationStatus = HostelVerificationStatus.Approved;
        hostel.UpdatedAt = DateTime.UtcNow;

        await _repo.SaveChangesAsync(cancellationToken);

        return _mapper.Map<HostelVerificationRequestReadDto>(request);
    }

    public async Task<HostelVerificationRequestReadDto> RejectVerificationAsync(Guid requestId, Guid adminId, string? adminNotes, CancellationToken cancellationToken)
    {
        var request = await _repo.GetRequestForUpdateAsync(requestId, cancellationToken)
            ?? throw new NotFoundException("Verification request not found.");

        if (request.Status != HostelVerificationStatus.Pending)
        {
            throw new BadRequestException("Only pending requests can be rejected.", "verification_request_not_pending");
        }

        request.Status = HostelVerificationStatus.Rejected;
        request.AdminNotes = string.IsNullOrWhiteSpace(adminNotes) ? null : adminNotes.Trim();
        request.ReviewedByAdminId = adminId;
        request.ReviewedAt = DateTime.UtcNow;
        request.UpdatedAt = DateTime.UtcNow;

        var hostel = await _repo.GetHostelForUpdateAsync(request.HostelId, cancellationToken)
            ?? throw new NotFoundException("Hostel not found.");

        hostel.IsVerified = false;
        hostel.VerifiedAt = null;
        hostel.VerifiedByAdminId = adminId;
        hostel.VerificationStatus = HostelVerificationStatus.Rejected;
        hostel.UpdatedAt = DateTime.UtcNow;

        await _repo.SaveChangesAsync(cancellationToken);

        return _mapper.Map<HostelVerificationRequestReadDto>(request);
    }

    public async Task<List<HostelVerificationRequestReadDto>> GetForHostelAsync(Guid hostelId, Guid requesterId, bool isAdmin, CancellationToken cancellationToken)
    {
        if (!isAdmin)
        {
            var hostel = await _repo.GetHostelForOwnerAsNoTrackingAsync(hostelId, requesterId, cancellationToken);
            if (hostel is null)
            {
                throw new ForbiddenException("You can only view verification requests for your own hostel.", "hostel_owner_required");
            }
        }

        var requests = await _repo.GetRequestsForHostelAsNoTrackingAsync(hostelId, cancellationToken);
        return _mapper.Map<List<HostelVerificationRequestReadDto>>(requests);
    }
}

File: src/hostels/verification/hostel-verification.service.ts
import { Injectable } from '@nestjs/common';

import { HostelVerificationStatus } from '../../common/enums/app.enums';
import {
  AppBadRequestException,
  AppConflictException,
  AppForbiddenException,
  AppNotFoundException,
} from '../../common/exceptions/app-exception';
import { DatabaseService } from '../../database/database.service';
import { HostelVerificationRequestReadDto } from './dto/hostel-verification.dto';

@Injectable()
export class HostelVerificationService {
  constructor(private readonly db: DatabaseService) {}

  async requestVerification(hostelId: string, ownerId: string): Promise<HostelVerificationRequestReadDto> {
    const hostel = await this.db.hostel.findFirst({
      where: {
        id: hostelId,
        ownerId,
        isDeleted: false,
      },
    });

    if (!hostel) {
      throw new AppForbiddenException('You can only request verification for your own hostel.', 'hostel_owner_required');
    }

    const pending = await this.db.hostelVerificationRequest.findFirst({
      where: {
        hostelId,
        status: HostelVerificationStatus.Pending,
        isDeleted: false,
      },
    });

    if (pending) {
      throw new AppConflictException('A pending verification request already exists.', 'verification_pending_exists');
    }

    const request = await this.db.$transaction(async (database) => {
      const created = await database.hostelVerificationRequest.create({
        data: {
          hostelId,
          requestedByUserId: ownerId,
          status: HostelVerificationStatus.Pending,
          createdAt: new Date(),
          isDeleted: false,
        },
      });

      await database.hostel.update({
        where: { id: hostelId },
        data: {
          verificationStatus: HostelVerificationStatus.Pending,
          updatedAt: new Date(),
        },
      });

      return created;
    });

    return this.toReadDto(request);
  }

  async approveVerification(
    requestId: string,
    adminId: string,
    adminNotes?: string | null,
  ): Promise<HostelVerificationRequestReadDto> {
    const request = await this.db.hostelVerificationRequest.findFirst({
      where: {
        id: requestId,
        isDeleted: false,
      },
    });

    if (!request) {
      throw new AppNotFoundException('Verification request not found.');
    }

    if (request.status !== HostelVerificationStatus.Pending) {
      throw new AppBadRequestException('Only pending requests can be approved.', 'verification_request_not_pending');
    }

    const updatedRequest = await this.db.$transaction(async (database) => {
      const approved = await database.hostelVerificationRequest.update({
        where: { id: requestId },
        data: {
          status: HostelVerificationStatus.Approved,
          adminNotes: adminNotes?.trim() || null,
          reviewedByAdminId: adminId,
          reviewedAt: new Date(),
          updatedAt: new Date(),
        },
      });

      await database.hostel.update({
        where: { id: request.hostelId },
        data: {
          isVerified: true,
          verifiedAt: new Date(),
          verifiedByAdminId: adminId,
          verificationStatus: HostelVerificationStatus.Approved,
          updatedAt: new Date(),
        },
      });

      return approved;
    });

    return this.toReadDto(updatedRequest);
  }

  async rejectVerification(
    requestId: string,
    adminId: string,
    adminNotes?: string | null,
  ): Promise<HostelVerificationRequestReadDto> {
    const request = await this.db.hostelVerificationRequest.findFirst({
      where: {
        id: requestId,
        isDeleted: false,
      },
    });

    if (!request) {
      throw new AppNotFoundException('Verification request not found.');
    }

    if (request.status !== HostelVerificationStatus.Pending) {
      throw new AppBadRequestException('Only pending requests can be rejected.', 'verification_request_not_pending');
    }

    const updatedRequest = await this.db.$transaction(async (database) => {
      const rejected = await database.hostelVerificationRequest.update({
        where: { id: requestId },
        data: {
          status: HostelVerificationStatus.Rejected,
          adminNotes: adminNotes?.trim() || null,
          reviewedByAdminId: adminId,
          reviewedAt: new Date(),
          updatedAt: new Date(),
        },
      });

      await database.hostel.update({
        where: { id: request.hostelId },
        data: {
          isVerified: false,
          verifiedAt: null,
          verifiedByAdminId: adminId,
          verificationStatus: HostelVerificationStatus.Rejected,
          updatedAt: new Date(),
        },
      });

      return rejected;
    });

    return this.toReadDto(updatedRequest);
  }

  async getForHostel(hostelId: string, requesterId: string, isAdmin: boolean): Promise<HostelVerificationRequestReadDto[]> {
    if (!isAdmin) {
      const hostel = await this.db.hostel.findFirst({
        where: {
          id: hostelId,
          ownerId: requesterId,
          isDeleted: false,
        },
      });

      if (!hostel) {
        throw new AppForbiddenException(
          'You can only view verification requests for your own hostel.',
          'hostel_owner_required',
        );
      }
    }

    const requests = await this.db.hostelVerificationRequest.findMany({
      where: {
        hostelId,
        isDeleted: false,
      },
      orderBy: { createdAt: 'desc' },
    });

    return requests.map((request) => this.toReadDto(request));
  }

  private toReadDto(request: any): HostelVerificationRequestReadDto {
    return {
      id: request.id,
      hostelId: request.hostelId,
      requestedByUserId: request.requestedByUserId,
      status: request.status,
      adminNotes: request.adminNotes,
      reviewedByAdminId: request.reviewedByAdminId,
      reviewedAt: request.reviewedAt,
      createdAt: request.createdAt,
    };
  }
}
